gglasso/solver/single_admm_solver.py

```diff
@@ -7,154 +7,199 @@
+        proxL = prox_rank_norm(A=L - X, beta=mu1, D=eigD, Q=eigQ)
+        term4 = np.linalg.norm(L - proxL) / (1 + np.linalg.norm(L))
+
+    criterion = np.array([])
+
+    #primal convergence condition
+    if r_k <= e_pri:
+        criterion = criterion.append("primal optimal")
+        print("primal problem has reached the optimal solution")
+
+    #dual convergence condition
+    if s_k <= e_dual:
+        criterion = criterion.append("dual optimal")
+        print("dual problem has reached the optimal solution")
+
+    stop_value = np.maximum(r_k, s_k, term4)
+
+    return {'status': criterion, 'value': stop_value}
+
+
+# def ADMM_stopping_criterion(Omega, Theta, L, X, S , lambda1, latent = False, mu1 = None):
+#
+#     assert Omega.shape == Theta.shape == S.shape
+#     assert S.shape[0] == S.shape[1]
+#
+#     if not latent:
+#         assert np.all(L==0)
+#
+#     (p,p) = S.shape
+#
+#     term1 = np.linalg.norm(Theta - prox_od_1norm(Theta + X , l = lambda1)) / (1 + np.linalg.norm(Theta))
+#
+#     term2 = np.linalg.norm(Omega - Theta + L) / (1 + np.linalg.norm(Theta))
+#
+#     eigD, eigQ = np.linalg.eigh(Omega - S - X)
+#     proxO = phiplus(A = Omega - S - X, beta = 1, D = eigD, Q = eigQ)
+#     term3 = np.linalg.norm(Omega - proxO) / (1 + np.linalg.norm(Omega))
+#
+#     term4 = 0
+#     if latent:
+#         eigD, eigQ = np.linalg.eigh(L - X)
+#         proxL = prox_rank_norm(A = L - X, beta = mu1, D = eigD, Q = eigQ)
+#         term4 = np.linalg.norm(L - proxL) / (1 + np.linalg.norm(L))
+#
+#     return max(term1, term2, term3, term4)
+
+
+def ADMM_SGL(S, lambda1, Omega_0, Theta_0=np.array([]), X_0=np.array([]), \
+             eps_admm=1e-5, rho=1., max_iter=1000, verbose=False, measure=False, latent=False, mu1=None):
     """
     This is an ADMM algorithm for solving the Single Graphical Lasso problem
-    
+
     Omega_0 : start point -- must be specified as a (p,p) array
     S : empirical covariance matrix -- must be specified as a (p,p) array
-    
+
     latent: boolean to indidate whether low rank term should be estimated
     mu1: low rank penalty paramater, if latent=True
-    
-    
+
+
     In the code, X are the SCALED (with 1/rho) dual variables, for the KKT stop criterion they have to be unscaled (i.e. take rho*X) again!
     """
     assert Omega_0.shape == S.shape
     assert S.shape[0] == S.shape[1]
     assert lambda1 > 0
-        
+
     if latent:
         assert mu1 is not None
         assert mu1 > 0
-        
-    (p,p) = S.shape
-    
+
+    (p, p) = S.shape
+
     assert rho > 0, "ADMM penalization parameter must be positive."
-        
-    
-    # initialize 
+
+    # initialize
     status = 'not optimal'
     Omega_t = Omega_0.copy()
     if len(Theta_0) == 0:
         Theta_0 = Omega_0.copy()
     if len(X_0) == 0:
-        X_0 = np.zeros((p,p))
+        X_0 = np.zeros((p, p))
 
     Theta_t = Theta_0.copy()
-    L_t = np.zeros((p,p))
+    Theta_t_1 = Theta_t
```

Comment:
`Theta_t_1` is the Theta-iterate from the previous iterate, so 1) in first iteration it is not correct and 2) it has to be updated also in the loop later on.
Have in mind that numpy will modify Theta_t_1 if you change Theta_t. In order to avoid this I always use `.copy`